services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/internal/DownloadDirectoryHelper.java

```diff
@@ -110,78 +112,167 @@ private void doDownloadDirectory(CompletableFuture<CompletedDirectoryDownload> r
-                                  .destination(destinationPath)
-                                  .getObjectRequest(getObjectRequest)
-                                  .build();
-    }
+        private CompletableFuture<CompletedFileDownload> downloadSingleFile(DownloadDirectoryRequest downloadDirectoryRequest,
+                                                                            Collection<FailedFileDownload> failedFileDownloads,
+                                                                            S3Object s3Object) {
+            FileSystem fileSystem = downloadDirectoryRequest.destinationDirectory().getFileSystem();
+            String delimiter = downloadDirectoryRequest.delimiter().orElse(DEFAULT_DELIMITER);
+
+            String key = normalizeKey(downloadDirectoryRequest, s3Object, delimiter);
+
+            String relativePath = getRelativePath(fileSystem,
+                                                  delimiter,
+                                                  key);
+
+            Path destinationPath = downloadDirectoryRequest.destinationDirectory().resolve(relativePath);
+            DownloadFileRequest downloadFileRequest = downloadFileRequest(downloadDirectoryRequest, s3Object, destinationPath);
+
+            try {
+                log.debug(() -> "Sending download request " + downloadFileRequest);
+
+                CompletableFuture<CompletedFileDownload> future =
+                    downloadFileFunction.apply(downloadFileRequest).completionFuture();
+                future.whenComplete((r, t) -> {
+                    if (t != null) {
+                        failedFileDownloads.add(FailedFileDownload.builder()
+                                                                  .exception(t)
+                                                                  .request(downloadFileRequest)
+                                                                  .build());
+                    }
+                });
+                return future;
+            } catch (Throwable throwable) {
+                failedFileDownloads.add(FailedFileDownload.builder()
+                                                          .exception(throwable)
+                                                          .request(downloadFileRequest)
+                                                          .build());
+                return CompletableFutureUtils.failedFuture(throwable);
+            }
+        }
+
+        /**
+         * Normalizing the key by stripping the prefix from the s3 object key if the prefix is not empty. For example: given a
+         * request with prefix = "notes/2021", delimiter = "/" and key = "notes/2021/1.txt", the returned string should be
+         * "1.txt"
+         */
+        private static String normalizeKey(DownloadDirectoryRequest downloadDirectoryRequest, S3Object s3Object,
+                                           String delimiter) {
+            return downloadDirectoryRequest.prefix()
```

Comment:
NonBlocker , just suggestion :
was curious if we can use regex for this , something like `^(.+)\/([^\/]+)$` where /  6th character is the delimiter
```
String regexExpression = String.format("^(.+)\\%s([^\\/]+)$", delimiter);
```

Reply:
I'm not sure if it's possible to define a regex because we don't know the length of the prefix beforehand.